proposals/filecoin-retrieval-in-ipfs.md

```diff
@@ -0,0 +1,163 @@
+
+- IPFS users would like to pin content and retrieve from Filecoin ideally without ever leaving IPFS or running any Filecoin software on their computer
+
+- The primary existing customers for retrieval of data ARE IPFS users. IPFS users don't want a seperate discovery mechanism for Filecoin retrieval -- they want to use what already exists
+
+- There is a market for paid retrieval within IPFS eventually and we should have a path to support this
+
+#### User workflow example
+_How would a developer or user use this new capability?_
+
+Example: IPFS pinning with Filecoin, using remote pinning API and free retrieval
+
+1. As an IPFS user who wants to pin data with filecoin, I negotiate an agreement to pin data in Filecoin with a pinning service out of band in exchange for an access token (i.e. I pay them through their website for X GB of filecoin pinning at Y Price). 
+
+2. I use the remote pinning API from the IPFS command line to pin some data with them.
+
+3. The pinning service initiates and completes a filecoin deal to store the data
+
+4. The pinning service, upon completion posts a record in the IPFS DHT for the CID of pinned data that contains both the peer address for the miner and also flag to indicate it requires filecoin retrieval (and potentially also includes the PieceCID?).
+
+5. For retrieval, upon querying the DHT, the IPFS client sees the the record is a filecoin retrieval record, and instead of using standard bitswap fetching, switches over to proposing a free retrieval via the data transfer protocol to the miner, and completes the retrieval this way.
+
+Future Example: Paid retrieval with IPFS & Lotus Lite client on same computer
+
+1. I make a deal with a miner through my lotus lite client to store some data. When my deal is done, the miner OR client can advertise the peer address and retrieval parameters in the IPFS dht 
+
+2. The IPFS Client sees a record for paid Filecoin retrieval in the DHT and uses the paid Filecoin retrieval plugin to talk to the Lotus Lite HTTP api to generate parameters for a paid retrieval deal. The IPFS client then initiates a data transfer request. As needed, the plugin calls out to Lotus Lite HTTP API to generate additional vouchers.
+
+Future example: Custom retrieval payment system for a "retrieval miner" using a custom payment system
+
+1. I negotiate out of band with a custom paid retrieval provider an agreement to retrieve data.
+
+2. I use the IPFS Pinning API to request to pin data. The miner stores the data in Filecoin as backup, and to prove it's being stored, but then distributes several copies to it's CDN pinning network and advertises all records in the IPFS DHT as paid retrievals that use its custom protocol.
+
+3. To retrieve, I use the information in the DHT and a plugin supplied by the provider to generate vouchers for a data transfer with one of their miners close to me. 
```

Comment:
Generally 👎 for plugins. They don't work on Windows and we have lots of users that use Windows. They can be useful for prototyping, and are fine whenever users are happy to build the binaries themselves, but otherwise not particularly good.

Reply:
I'm less worried about plugin vs rest vs ipc as long as there's a clear interface :)